src/ILCompiler.Compiler/src/IL/ILImporter.Scanner.cs

```diff
@@ -111,8 +111,13 @@ public DependencyList Import()
                 if (!_isFallbackBodyCompilation &&
                     (_canonMethod.Signature.IsStatic || _canonMethod.IsConstructor || owningType.IsValueType))
                 {
-                    // For beforefieldinit, we can wait for field access.
-                    if (!((MetadataType)owningType).IsBeforeFieldInit)
+                    // For beforefieldinit, we can wait for field access...
+                    bool runCctor = !((MetadataType)owningType).IsBeforeFieldInit;
+
+                    // ...unless this is a PInvoke, in which case we run it regardless
```

Comment:
> (when the program is compiled `/o+`).

Mmm, interesting. The purpose of the scanner is to collect the dependencies of the app without optimizations, so simulating that we need the cctor is still the correct thing to do. Without `/o+`, the cctor is triggered on desktop CLR too (which is what I tested with).

I can retest what RyuJIT will do with the CoreRT compiler with optimizations enabled later, but that is largely orthogonal to this change.